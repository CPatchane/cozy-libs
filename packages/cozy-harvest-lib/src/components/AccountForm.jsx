import React, { PureComponent } from 'react'
import { Form, Field as FinalFormField } from 'react-final-form'
import PropTypes from 'react-proptypes'

import Button from 'cozy-ui/react/Button'
import { translate, extend } from 'cozy-ui/react/I18n'
import Field from 'cozy-ui/react/Field'

import Manifest from '../Manifest'
import OAuthForm from './OAuthForm'

const ENCRYPTED_PLACEHOLDER = '*************'

const predefinedLabels = [
  'answer',
  'birthdate',
  'code',
  'date',
  'email',
  'firstname',
  'lastname',
  'login',
  'password',
  'phone'
]

export class AccountField extends PureComponent {
  render() {
    const { label, name, t, type } = this.props

    // Allow manifest to specify predefined label
    const localeKey = predefinedLabels.includes(label) ? label : name

    const fieldProps = {
      ...this.props,
      className: 'u-m-0', // 0 margin
      fullwidth: true,
      label: t(`fields.${localeKey}.label`, { _: name }),
      size: 'medium'
    }
    const passwordLabels = {
      hideLabel: t('accountForm.password.hide'),
      showLabel: t('accountForm.password.show')
    }
    switch (type) {
      case 'dropdown':
        return (
          <Field
            {...fieldProps}
            options={fieldProps.options.map(option => ({
              ...option,
              // legacy
              label: option.name
            }))}
            value={fieldProps.options.find(o => o.value === fieldProps.value)}
            type="select"
          />
        )
      case 'password':
        return <Field {...fieldProps} secondaryLabels={passwordLabels} />
      default:
        return <Field {...fieldProps} type="text" />
    }
  }
}

// As SelectBox component from Cozy-UI, rendering dropdown type, is just giving
// us the full Option object, we just get its value to facilitate mapping
// with account
const parse = type => value => {
  return type === 'dropdown' ? value.value : value
}

export class AccountFields extends PureComponent {
  render() {
    const { fillEncrypted, manifestFields, t } = this.props

    // Ready to use named fields array
    const namedFields = Object.keys(manifestFields).map(fieldName => ({
      ...manifestFields[fieldName],
      name: fieldName
    }))

    return (
      <div>
        {namedFields.map((field, index) => (
          <FinalFormField
            key={index}
            name={field.name}
            parse={parse(field.type)}
          >
            {({ input }) => (
              <AccountField
                {...field}
                {...input}
                placeholder={
                  field.encrypted && fillEncrypted
                    ? ENCRYPTED_PLACEHOLDER
                    : field.placeholder
                }
                t={t}
              />
            )}
          </FinalFormField>
        ))}
      </div>
    )
  }
}

export class AccountForm extends PureComponent {
  constructor(props, context) {
    super(props, context)
    const { lang, locales } = props
    if (locales && lang) {
      extend(locales[lang])
    }
  }

  render() {
    const { account, fields, oauth, t } = this.props

    if (oauth) return <OAuthForm oauth={oauth} />

    const sanitizedFields = Manifest.sanitizeFields(fields)
    const initialValues = account ? account.auth : {}
    return (
      <Form
        initialValues={initialValues}
        // eslint-disable-next-line no-console
        onSubmit={v => console.log(v)}
        render={({ values }) => (
          <div>
            <AccountFields
              fillEncrypted={!!account}
              manifestFields={sanitizedFields}
              t={t}
            />
            <Button
              className="u-mt-2 u-mb-1-half"
              extension="full"
              label={t('accountForm.submit.label')}
              onclick={() => alert(JSON.stringify(values, 0, 2))}
            />
          </div>
        )}
      />
    )
  }
}

AccountForm.propTypes = {
  account: PropTypes.object,
  fields: PropTypes.object.isRequired,
  locales: PropTypes.object
}

export default translate()(AccountForm)
