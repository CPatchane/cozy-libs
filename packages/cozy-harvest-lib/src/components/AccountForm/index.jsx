import React, { PureComponent } from 'react'
import { Form, Field as FinalFormField } from 'react-final-form'
import PropTypes from 'react-proptypes'

import Button from 'cozy-ui/react/Button'
import { translate, extend } from 'cozy-ui/react/I18n'

import AccountField from './AccountField'
import AccountFormError from './Error'
import { getEncryptedFieldName } from '../../helpers/fields'
import { KonnectorJobError } from 'helpers/konnectors'
import Manifest from 'Manifest'
import OAuthForm from 'components/OAuthForm'

const VALIDATION_ERROR_REQUIRED_FIELD = 'VALIDATION_ERROR_REQUIRED_FIELD'

// As SelectBox component from Cozy-UI, rendering dropdown type, is just giving
// us the full Option object, we just get its value to facilitate mapping
// with account
const parse = type => value => {
  return type === 'dropdown' ? value.value : value
}

export class AccountFields extends PureComponent {
  render() {
    const {
      container,
      disabled,
      hasError,
      initialValues,
      manifestFields,
      onKeyUp,
      t
    } = this.props

    // Ready to use named fields array
    const namedFields = Object.keys(manifestFields).map(fieldName => ({
      ...manifestFields[fieldName],
      name: fieldName
    }))

    return (
      <div onKeyUp={onKeyUp}>
        {namedFields.map((field, index) => (
          <FinalFormField
            key={index}
            name={field.name}
            parse={parse(field.type)}
          >
            {({ input }) => (
              <AccountField
                {...field}
                {...input}
                container={container}
                hasError={hasError}
                disabled={disabled}
                initialValue={
                  initialValues[field.name] ||
                  initialValues[getEncryptedFieldName(field.name)]
                }
                t={t}
              />
            )}
          </FinalFormField>
        ))}
      </div>
    )
  }
}

AccountFields.propTypes = {
  disabled: PropTypes.bool,
  hasError: PropTypes.bool,
  fillEncrypted: PropTypes.bool,
  manifestFields: PropTypes.object.isRequired,
  onKeyUp: PropTypes.func,
  t: PropTypes.func
}

export class AccountForm extends PureComponent {
  constructor(props, context) {
    super(props, context)
    const { konnector, lang } = props
    const { locales } = konnector
    if (locales && lang) {
      extend(locales[lang])
    }
  }

  /**
   * Indicates if the state of ReactFinalForm implies that data can be submitted
   * @param  {Object} formState See https://github.com/final-form/final-form#formstate
   * @return {Boolean}
   */
  isSubmittable({ dirty, error, initialValues, valid }) {
    const untouched = initialValues && !dirty
    return error || (valid && !untouched)
  }

  handleKeyUp(event, { dirty, form, initialValues, valid, values }) {
    if (
      event.code === 'Enter' &&
      this.isSubmittable({ dirty, initialValues, valid })
    ) {
      this.handleSubmit(values, form)
    }
  }

  handleSubmit(values, form) {
    const { onSubmit } = this.props
    // Reset form with new values to set back dirty to false
    form.reset(values)
    onSubmit(values)
  }

  validate = (fields, initialValues) => vals => {
    let errors = {}
    for (let name in fields)
      if (
        fields[name].required &&
        !vals[name] &&
        // Don't require value for empty encrypted fields with initial value
        !initialValues[getEncryptedFieldName(name)]
      )
        errors[name] = VALIDATION_ERROR_REQUIRED_FIELD
    return errors
  }

  render() {
    const {
      account,
      error,
      konnector,
      onSubmit,
      showError,
      submitting,
      t
    } = this.props
    const { fields, oauth } = konnector

    if (oauth) return <OAuthForm initialValues={initialValues} oauth={oauth} />

    const sanitizedFields = Manifest.sanitizeFields(fields)
    const defaultValues = Manifest.defaultFieldsValues(sanitizedFields)
    const initialValues = account && account.auth
    const initialAndDefaultValues = { ...defaultValues, ...initialValues }

    let container = null

    const isLoginError =
      error instanceof KonnectorJobError && error.isLoginError()

    return (
      <Form
        initialValues={initialAndDefaultValues}
        onSubmit={onSubmit}
        validate={this.validate(sanitizedFields, initialAndDefaultValues)}
        render={({ dirty, form, values, valid }) => (
          <div
            ref={element => {
              container = element
            }}
          >
            {error && (showError || isLoginError) && (
              <AccountFormError error={error} konnector={konnector} t={t} />
            )}
            <AccountFields
              container={container}
              disabled={submitting}
              hasError={
                error &&
                error instanceof KonnectorJobError &&
                error.isLoginError()
              }
              initialValues={initialAndDefaultValues}
              manifestFields={sanitizedFields}
              onKeyUp={event =>
                this.handleKeyUp(event, {
                  dirty,
                  form,
                  initialValues,
                  valid,
                  values
                })
              }
              t={t}
            />
            <Button
              busy={submitting}
              className="u-mt-2 u-mb-1-half"
              disabled={
                submitting ||
                !this.isSubmittable({ dirty, error, initialValues, valid })
              }
              extension="full"
              label={t('accountForm.submit.label')}
              onClick={() => this.handleSubmit(values, form)}
            />
          </div>
        )}
      />
    )
  }
}

AccountForm.propTypes = {
  account: PropTypes.object,
  konnector: PropTypes.object.isRequired,
  error: PropTypes.object,
  showError: PropTypes.bool,
  submitting: PropTypes.bool
}

AccountForm.defaultProps = {
  showError: true
}

export default translate()(AccountForm)
