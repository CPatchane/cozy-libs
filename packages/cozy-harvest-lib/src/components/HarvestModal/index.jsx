import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import Modal, { ModalDescription, ModalHeader } from 'cozy-ui/react/Modal'
import AppIcon from 'cozy-ui/react/AppIcon'
import Spinner from 'cozy-ui/react/Spinner'

import HarvestModalContent from './Content'
import withAccountFromTrigger from '../HOCs/withAccountFromTrigger'

export class HarvestModal extends PureComponent {
  constructor(props) {
    super(props)
    this.fetchIcon = this.fetchIcon.bind(this)
  }

  fetchIcon() {
    const { client } = this.context
    const { konnector } = this.props
    return client.stackClient.getIconURL({
      type: 'konnector',
      slug: konnector.slug
    })
  }

  render() {
    const {
      account,
      dismissAction,
      fetchingAccount,
      konnector,
      into,
      trigger,
      onLoginSuccess,
      onSuccess,
      running
    } = this.props

    return (
      <Modal
        dismissAction={dismissAction}
        mobileFullscreen
        size="small"
        into={into}
      >
        <ModalHeader>
          <AppIcon fetchIcon={this.fetchIcon} className="u-mah-3 u-ml-1" />
        </ModalHeader>
        <ModalDescription>
          {fetchingAccount ? (
            <Spinner
              size="xxlarge"
              className="u-flex u-flex-justify-center u-pv-3"
            />
          ) : (
            <HarvestModalContent
              account={account}
              konnector={konnector}
              showError={false}
              trigger={trigger}
              running={running}
              onLoginSuccess={onLoginSuccess}
              onSuccess={onSuccess}
            />
          )}
        </ModalDescription>
      </Modal>
    )
  }
}

HarvestModal.contextTypes = {
  client: PropTypes.object.isRequired
}

export default withAccountFromTrigger(HarvestModal)
