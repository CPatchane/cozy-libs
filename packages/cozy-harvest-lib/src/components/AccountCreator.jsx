import React, { PureComponent } from 'react'
import PropTypes from 'react-proptypes'

import { withMutations } from 'cozy-client'

import AccountForm from './AccountForm'
import { accountsMutations } from '../connections/accounts'
import { mergeAuth, prepareAccountData } from '../helpers/accounts'

/**
 * Encapsulates an AccountForm and create an account with resulting data.
 */
export class AccountCreator extends PureComponent {
  constructor(props) {
    super(props)
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  async handleSubmit(data) {
    const {
      createAccount,
      konnector,
      onBeforeCreate,
      onCreateSuccess
    } = this.props
    onBeforeCreate()
    const account = await createAccount(
      konnector,
      prepareAccountData(konnector, data)
    )
    // Merge auth to keep original values for encrypted fields during creation
    onCreateSuccess(mergeAuth(account, data))
  }

  render() {
    const { account, konnector, submitting } = this.props
    return (
      <AccountForm
        fields={konnector.fields}
        initialValues={account && (account.auth || account.oauth)}
        locales={konnector.locales}
        oauth={konnector.oauth}
        onSubmit={this.handleSubmit}
        submitting={submitting}
      />
    )
  }
}

AccountCreator.propTypes = {
  createAccount: PropTypes.func.isRequired,
  konnector: PropTypes.object.isRequired,
  submitting: PropTypes.bool,
  onBeforeCreate: PropTypes.func.isRequired,
  onCreateSuccess: PropTypes.func.isRequired
}

export default withMutations(accountsMutations)(AccountCreator)
