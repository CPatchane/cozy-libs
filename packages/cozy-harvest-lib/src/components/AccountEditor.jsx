import React, { PureComponent } from 'react'
import PropTypes from 'react-proptypes'

import { withMutations } from 'cozy-client'

import AccountForm from './AccountForm'
import { accountsMutations } from '../connections/accounts'
import { mergeAuth } from '../helpers/accounts'

/**
 * Encapsulates an AccountForm of an existing account, and allows to update it.
 */
export class AccountEditor extends PureComponent {
  constructor(props) {
    super(props)
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  async handleSubmit(data) {
    const {
      account,
      onBeforeUpdate,
      onUpdateSuccess,
      updateAccount
    } = this.props
    onBeforeUpdate(account)
    const updatedAccount = await updateAccount(mergeAuth(account, data))
    onUpdateSuccess(updatedAccount)
  }

  render() {
    const { account, konnector, submitting } = this.props
    return (
      <AccountForm
        fields={konnector.fields}
        initialValues={account && (account.auth || account.oauth)}
        locales={konnector.locales}
        oauth={konnector.oauth}
        onSubmit={this.handleSubmit}
        submitting={submitting}
      />
    )
  }
}

AccountEditor.propTypes = {
  account: PropTypes.object.isRequired,
  konnector: PropTypes.object.isRequired,
  onBeforeUpdate: PropTypes.func.isRequired,
  onUpdateSuccess: PropTypes.func.isRequired,
  submitting: PropTypes.bool,
  updateAccount: PropTypes.func.isRequired
}

export default withMutations(accountsMutations)(AccountEditor)
