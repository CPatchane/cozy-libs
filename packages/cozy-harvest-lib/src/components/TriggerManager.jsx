import React, { Component } from 'react'
import PropTypes from 'react-proptypes'

import { withMutations } from 'cozy-client'
import { translate } from 'cozy-ui/react/I18n'

import AccountForm from './AccountForm'
import accountsMutations from '../connections/accounts'
import { triggersMutations } from '../connections/triggers'
import filesMutations from '../connections/files'
import permissionsMutations from '../connections/permissions'
import accounts from '../helpers/accounts'
import cron from '../helpers/cron'
import konnectors from '../helpers/konnectors'
import { slugify } from '../helpers/slug'
import triggers from '../helpers/triggers'

const IDLE = 'IDLE'
const RUNNING = 'RUNNING'

/**
 * Deals with konnector configuration, i.e encapsulate account creation or
 * edition
 * @type {Component}
 */
export class TriggerManager extends Component {
  state = {
    account: null,
    status: IDLE
  }

  constructor(props) {
    super(props)

    this.handleAccountCreationSuccess = this.handleAccountCreationSuccess.bind(
      this
    )
    this.handleAccountUpdateSuccess = this.handleAccountUpdateSuccess.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)

    this.state = {
      account: props.account,
      status: IDLE
    }
  }

  /**
   * Account creation success handler
   * @param  {Object}  account Created io.cozy.accounts document
   * @return {Object}          io.cozy.jobs document, runned with account data
   */
  async handleAccountCreationSuccess(account) {
    const {
      addPermission,
      addReferencesTo,
      createDirectoryByPath,
      createTrigger,
      statDirectoryByPath,
      konnector,
      t
    } = this.props

    this.setState({
      account
    })

    let folder
    if (konnectors.needsFolder(konnector)) {
      const path = `${t('default.baseDir')}/${konnector.name}/${slugify(
        accounts.getLabel(account)
      )}`

      folder =
        (await statDirectoryByPath(path)) || (await createDirectoryByPath(path))

      await addPermission(konnector, konnectors.buildFolderPermission(folder))
      await addReferencesTo(konnector, [folder])
    }

    const trigger = await createTrigger(
      triggers.buildAttributes({
        account,
        cron: cron.fromKonnector(konnector),
        folder,
        konnector
      })
    )

    return await this.launch(trigger)
  }

  /**
   * Account update success handler
   * @param  {Object}  account Updated io.cozy.accounts document
   * @return {Object}          io.cozy.jobs document, runned with account data
   */
  async handleAccountUpdateSuccess(account) {
    this.setState({
      account
    })

    const { trigger } = this.props
    return await this.launch(trigger)
  }

  async handleSubmit(data) {
    const { konnector, saveAccount } = this.props

    const { account } = this.state
    const isUpdate = !!account

    this.setState({
      status: RUNNING
    })

    const savedAccount = accounts.mergeAuth(
      await saveAccount(
        konnector,
        isUpdate
          ? accounts.mergeAuth(account, data)
          : accounts.build(konnector, data)
      ),
      data
    )

    return isUpdate
      ? this.handleAccountUpdateSuccess(savedAccount)
      : this.handleAccountCreationSuccess(savedAccount)
  }

  /**
   * Launches a trigger
   * @param  {Object}  trigger io.cozy.triggers document
   * @return {Promise}         [description]
   */
  async launch(trigger) {
    const {
      launchTrigger,
      onLoginSuccess,
      onSuccess,
      waitForLoginSuccess
    } = this.props

    const job = await waitForLoginSuccess(await launchTrigger(trigger))

    this.setState({
      status: IDLE
    })

    if (['queued', 'running'].includes(job.state)) {
      return onLoginSuccess(trigger)
    }

    return onSuccess(trigger)
  }

  render() {
    const { konnector, running } = this.props
    const { account, error, status } = this.state
    const submitting = status === RUNNING || running

    return (
      <AccountForm
        account={account}
        error={error}
        konnector={konnector}
        onSubmit={this.handleSubmit}
        submitting={submitting}
      />
    )
  }
}

TriggerManager.propTypes = {
  account: PropTypes.object,
  konnector: PropTypes.object.isRequired,
  trigger: PropTypes.object,
  running: PropTypes.bool,
  // mutations
  addPermission: PropTypes.func,
  addReferencesTo: PropTypes.func,
  createTrigger: PropTypes.func.isRequired,
  createDirectoryByPath: PropTypes.func,
  launchTrigger: PropTypes.func.isRequired,
  saveAccount: PropTypes.func.isRequired,
  statDirectoryByPath: PropTypes.func,
  waitForLoginSuccess: PropTypes.func.isRequired,
  // hooks
  onLoginSuccess: PropTypes.func.isRequired,
  onSuccess: PropTypes.func
}

export default translate()(
  withMutations(
    accountsMutations,
    filesMutations,
    permissionsMutations,
    triggersMutations
  )(TriggerManager)
)
