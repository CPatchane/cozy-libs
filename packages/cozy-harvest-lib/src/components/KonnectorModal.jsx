import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import { withMutations } from 'cozy-client'
import AppIcon from 'cozy-ui/react/AppIcon'
import Modal, { ModalContent, ModalHeader } from 'cozy-ui/react/Modal'
import Spinner from 'cozy-ui/react/Spinner'
import { SubTitle } from 'cozy-ui/react/Text'

import accountMutations from '../connections/accounts'
import * as triggers from '../helpers/triggers'
import TriggerManager from './TriggerManager'
import withLocales from './hoc/withLocales'

/**
 * KonnectorModal open a Modal related to a given konnector. It fetches the
 * first account and then include a TriggerManager component.
 *
 * This component is aimed to offer an UI to manage all the konnector related
 * triggers and accounts.
 */
export class KonnectorModal extends PureComponent {
  state = {
    account: null,
    fetching: true,
    trigger: null
  }

  constructor(props) {
    super(props)
    this.fetchIcon = this.fetchIcon.bind(this)
  }

  componentDidMount() {
    this.fetchAccount()
  }

  async fetchAccount() {
    const { findAccount } = this.props
    this.setState({ fetching: true })
    const { konnector } = this.props
    const trigger = konnector.triggers.data[0]
    const account = await findAccount(triggers.getAccountId(trigger))
    this.setState({
      account,
      fetching: false
    })
  }

  fetchIcon() {
    const { client } = this.context
    const { konnector } = this.props
    return client.stackClient.getIconURL({
      type: 'konnector',
      slug: konnector.slug
    })
  }

  render() {
    const {
      dismissAction,
      konnector,
      into,
      onLoginSuccess,
      onSuccess,
      running,
      t
    } = this.props

    const { account, fetching, trigger } = this.state

    return (
      <Modal
        dismissAction={dismissAction}
        mobileFullscreen
        size="small"
        into={into}
      >
        <ModalHeader>
          <AppIcon fetchIcon={this.fetchIcon} className="u-mah-3 u-ml-1" />
        </ModalHeader>
        {fetching ? (
          <ModalContent>
            <Spinner
              size="xxlarge"
              className="u-flex u-flex-justify-center u-pv-3"
            />
          </ModalContent>
        ) : (
          <ModalContent>
            <SubTitle className="u-mb-1 u-ta-center">
              {t('modal.konnector.title', { name: konnector.name })}
            </SubTitle>
            {error ? (
              <Infos
                actionButton={
                  <Button theme="danger">
                    {t('modal.konnector.error.button')}
                  </Button>
                }
                title={t('modal.konnector.error.title')}
                text={t('modal.konnector.error.description', error)}
                isImportant
              />
            ) : (
              <TriggerManager
                account={account}
                konnector={konnector}
                trigger={trigger}
                running={running}
                onLoginSuccess={onLoginSuccess}
                onSuccess={onSuccess}
              />
            )}
          </ModalContent>
        )}
      </Modal>
    )
  }
}

KonnectorModal.contextTypes = {
  client: PropTypes.object.isRequired
}

export default withMutations(accountMutations)(withLocales(KonnectorModal))
