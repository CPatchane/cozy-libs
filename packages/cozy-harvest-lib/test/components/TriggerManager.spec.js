/* eslint-env jest */
import React from 'react'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import { TriggerManager } from 'components/TriggerManager'
import cronHelpers from 'helpers/cron'

configure({ adapter: new Adapter() })

const fixtures = {
  data: {
    username: 'foo',
    passphrase: 'bar'
  },
  konnector: {
    slug: 'konnectest'
  },
  konnectorWithFolder: {
    name: 'myBills',
    slug: 'mybills',
    fields: {
      advancedFields: {
        folderPath: {
          advanced: true
        }
      }
    }
  },
  folder: {
    _id: '3f5b288af36041f189ec22063adab706'
  },
  folderPath: '/Administrative/myBills/foo',
  folderPermission: {
    saveFolder: {
      type: 'io.cozy.files',
      values: ['3f5b288af36041f189ec22063adab706'],
      verbs: ['GET', 'PATCH', 'POST']
    }
  },
  triggerAttributes: {
    arguments: '0 0 0 * * 0',
    type: '@cron',
    worker: 'konnector',
    message: {
      account: 'a87f9a8bd3884479a48811e7b7deec75',
      konnector: 'konnectest'
    }
  },
  account: {
    _id: 'a87f9a8bd3884479a48811e7b7deec75',
    account_type: 'konnectest',
    auth: {
      username: 'foo',
      passphrase: 'bar'
    },
    identifier: 'username'
  },
  updatedAccount: {
    _id: 'a87f9a8bd3884479a48811e7b7deec75',
    account_type: 'konnectest',
    auth: {
      username: 'foo',
      passphrase: 'fuz'
    }
  },
  createdTrigger: {
    id: '669e9a7cc3064a97bc0aa20feef71cb2',
    _type: 'io.cozy.triggers',
    attributes: {
      arguments: '0 0 0 * * 0',
      type: '@cron',
      worker: 'konnector',
      message: {
        account: 'a87f9a8bd3884479a48811e7b7deec75',
        konnector: 'konnectest'
      }
    }
  },
  launchedJob: {
    type: 'io.cozy.jobs',
    id: 'ac09e6f4473f4b6fbb83c9d2f532504e',
    domain: 'cozy.tools:8080',
    worker: 'konnector',
    state: 'running',
    queued_at: '2016-09-19T12:35:08Z',
    started_at: '2016-09-19T12:35:08Z',
    error: '',
    links: {
      self: '/jobs/ac09e6f4473f4b6fbb83c9d2f532504e'
    }
  },
  runningJob: {
    type: 'io.cozy.jobs',
    id: 'ac09e6f4473f4b6fbb83c9d2f532504e',
    domain: 'cozy.tools:8080',
    worker: 'konnector',
    state: 'running',
    queued_at: '2016-09-19T12:35:08Z',
    started_at: '2016-09-19T12:35:08Z',
    error: '',
    links: {
      self: '/jobs/ac09e6f4473f4b6fbb83c9d2f532504e'
    }
  },
  doneJob: {
    type: 'io.cozy.jobs',
    id: 'ac09e6f4473f4b6fbb83c9d2f532504e',
    domain: 'cozy.tools:8080',
    worker: 'konnector',
    state: 'done',
    queued_at: '2016-09-19T12:35:08Z',
    started_at: '2016-09-19T12:35:08Z',
    error: '',
    links: {
      self: '/jobs/ac09e6f4473f4b6fbb83c9d2f532504e'
    }
  }
}

const addPermissionMock = jest.fn()
const addReferencesToMock = jest.fn()
const createTriggerMock = jest.fn().mockResolvedValue(fixtures.createdTrigger)
const createDirectoryByPathMock = jest.fn()
const statDirectoryByPathMock = jest.fn()
const launchTriggerMock = jest.fn().mockResolvedValue(fixtures.launchedJob)
const waitForLoginSuccessMock = jest.fn().mockResolvedValue(fixtures.runningJob)

const onDoneSpy = jest.fn()
const onLoginSuccessSpy = jest.fn()

const tMock = jest.fn().mockReturnValue('/Administrative')

const shallowAccountCreator = konnector =>
  shallow(
    <TriggerManager
      addPermission={addPermissionMock}
      addReferencesTo={addReferencesToMock}
      konnector={konnector || fixtures.konnector}
      createTrigger={createTriggerMock}
      createDirectoryByPath={createDirectoryByPathMock}
      statDirectoryByPath={statDirectoryByPathMock}
      launchTrigger={launchTriggerMock}
      onDone={onDoneSpy}
      onLoginSuccess={onLoginSuccessSpy}
      t={tMock}
      waitForLoginSuccess={waitForLoginSuccessMock}
    />
  )

const shallowAccountEditor = () =>
  shallow(
    <TriggerManager
      account={fixtures.account}
      createTrigger={createTriggerMock}
      konnector={fixtures.konnector}
      launchTrigger={launchTriggerMock}
      onDone={onDoneSpy}
      onLoginSuccess={onLoginSuccessSpy}
      trigger={fixtures.createdTrigger}
      waitForLoginSuccess={waitForLoginSuccessMock}
    />
  )

describe('TriggerManager', () => {
  afterEach(() => {
    createTriggerMock.mockClear()
    launchTriggerMock.mockClear()
    onDoneSpy.mockClear()
    onLoginSuccessSpy.mockClear()
    waitForLoginSuccessMock.mockClear()
  })

  it('should render AccountCreator', () => {
    const component = shallowAccountCreator().getElement()
    expect(component).toMatchSnapshot()
  })

  it('should render AccountEditor', () => {
    const component = shallowAccountEditor().getElement()
    expect(component).toMatchSnapshot()
  })

  describe('handleAccountMutation', () => {
    it('should render AccountCreator as submitting', () => {
      const wrapper = shallowAccountCreator()
      wrapper.instance().handleAccountMutation()
      expect(wrapper.props().submitting).toEqual(true)
    })

    it('should render AccountCreator as submitting', () => {
      const wrapper = shallowAccountEditor()
      wrapper.instance().handleAccountMutation()
      expect(wrapper.props().submitting).toEqual(true)
    })
  })

  describe('handleAccountCreationSuccess', () => {
    beforeAll(() => {
      jest.spyOn(cronHelpers, 'fromFrequency').mockReturnValue('0 0 0 * * 0')
    })

    afterEach(() => {
      jest.clearAllMocks()
    })

    afterAll(() => {
      jest.restoreAllMocks()
    })

    it('should create trigger', async () => {
      const wrapper = shallowAccountCreator()
      await wrapper.instance().handleAccountCreationSuccess(fixtures.account)
      expect(createTriggerMock).toHaveBeenCalledTimes(1)
      expect(createTriggerMock).toHaveBeenCalledWith(fixtures.triggerAttributes)
    })

    it('should launch trigger', async () => {
      const wrapper = shallowAccountCreator()
      await wrapper.instance().handleAccountCreationSuccess(fixtures.account)
      expect(launchTriggerMock).toHaveBeenCalledTimes(1)
      expect(launchTriggerMock).toHaveBeenCalledWith(fixtures.createdTrigger)
    })

    it('should render TriggerSuccessMessage', async () => {
      const wrapper = shallowAccountCreator()
      await wrapper.instance().handleAccountCreationSuccess(fixtures.account)
      const component = wrapper.getElement()
      expect(component).toMatchSnapshot()
    })

    describe('when konnector needs folder', () => {
      it('should create folder if it does not exist', async () => {
        statDirectoryByPathMock.mockResolvedValue(null)
        createDirectoryByPathMock.mockReturnValue(fixtures.folder)

        const wrapper = shallowAccountCreator(fixtures.konnectorWithFolder)
        await wrapper.instance().handleAccountCreationSuccess(fixtures.account)

        expect(statDirectoryByPathMock).toHaveBeenCalledTimes(1)
        expect(createDirectoryByPathMock).toHaveBeenCalledTimes(1)
        expect(createDirectoryByPathMock).toHaveBeenCalledWith(
          fixtures.folderPath
        )
        expect(addPermissionMock).toHaveBeenCalledTimes(1)
        expect(addPermissionMock).toHaveBeenCalledWith(
          fixtures.konnectorWithFolder,
          fixtures.folderPermission
        )
        expect(addReferencesToMock).toHaveBeenCalledTimes(1)
        expect(addReferencesToMock).toHaveBeenCalledWith(
          fixtures.konnectorWithFolder,
          [fixtures.folder]
        )
      })

      it('should not create folder if it exists', async () => {
        statDirectoryByPathMock.mockResolvedValue(fixtures.folder)

        const wrapper = shallowAccountCreator(fixtures.konnectorWithFolder)
        await wrapper.instance().handleAccountCreationSuccess(fixtures.account)

        expect(statDirectoryByPathMock).toHaveBeenCalledTimes(1)
        expect(createDirectoryByPathMock).toHaveBeenCalledTimes(0)

        expect(addPermissionMock).toHaveBeenCalledTimes(1)
        expect(addPermissionMock).toHaveBeenCalledWith(
          fixtures.konnectorWithFolder,
          fixtures.folderPermission
        )
        expect(addReferencesToMock).toHaveBeenCalledTimes(1)
        expect(addReferencesToMock).toHaveBeenCalledWith(
          fixtures.konnectorWithFolder,
          [fixtures.folder]
        )
      })
    })
  })

  describe('handleAccountUpdateSuccess', () => {
    it('should launch trigger', async () => {
      const wrapper = shallowAccountEditor()
      await wrapper
        .instance()
        .handleAccountUpdateSuccess(fixtures.updatedAccount)
      expect(launchTriggerMock).toHaveBeenCalledTimes(1)
      expect(launchTriggerMock).toHaveBeenCalledWith(fixtures.createdTrigger)
    })

    it('should render AccountEditor with updated account', async () => {
      const wrapper = shallowAccountEditor()
      await wrapper
        .instance()
        .handleAccountUpdateSuccess(fixtures.updatedAccount)
      expect(wrapper.props().account).toEqual(fixtures.updatedAccount)
    })
  })

  describe('launch', () => {
    it('should launch trigger', async () => {
      const wrapper = shallowAccountCreator()
      await wrapper.instance().launch(fixtures.createdTrigger)
      expect(launchTriggerMock).toHaveBeenCalledTimes(1)
      expect(launchTriggerMock).toHaveBeenCalledWith(fixtures.createdTrigger)
    })

    it('should wait for successful login', async () => {
      const wrapper = shallowAccountCreator()
      await wrapper.instance().launch(fixtures.createdTrigger)
      expect(waitForLoginSuccessMock).toHaveBeenCalledTimes(1)
      expect(waitForLoginSuccessMock).toHaveBeenCalledWith(fixtures.launchedJob)
    })

    it('should call onLoginSuccess', async () => {
      const wrapper = shallowAccountCreator()
      await wrapper.instance().launch(fixtures.createdTrigger)
      expect(onLoginSuccessSpy).toHaveBeenCalledTimes(1)
      expect(onLoginSuccessSpy).toHaveBeenCalledWith(fixtures.createdTrigger)
    })

    it('should not call onLoginSucces if job is done', async () => {
      waitForLoginSuccessMock.mockResolvedValue(fixtures.doneJob)
      const wrapper = shallowAccountCreator()
      await wrapper.instance().launch(fixtures.createdTrigger)
      expect(onLoginSuccessSpy).not.toHaveBeenCalled()
      waitForLoginSuccessMock.mockResolvedValue(fixtures.runningJob)
    })
  })
})
