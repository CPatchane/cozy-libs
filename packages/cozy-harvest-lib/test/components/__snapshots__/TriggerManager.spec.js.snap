// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TriggerManager should render with account 1`] = `
<Wrapper
  account={
    Object {
      "_id": "a87f9a8bd3884479a48811e7b7deec75",
      "account_type": "konnectest",
      "auth": Object {
        "passphrase": "bar",
        "username": "foo",
      },
      "identifier": "username",
    }
  }
  konnector={
    Object {
      "slug": "konnectest",
    }
  }
  onSubmit={[Function]}
/>
`;

exports[`TriggerManager should render without account 1`] = `
<Wrapper
  konnector={
    Object {
      "slug": "konnectest",
    }
  }
  onSubmit={[Function]}
/>
`;
